### PR TITLE
issue-444:Convert Pytest-CVP Fan Status test case to Vane 

### PR DESCRIPTION
**Pull request by rohitsinghgslab**
# Please include a summary of the changes

Converted test_hardware_fan_status.py as per the vane style guide along with that updated following files.
test_definition.yaml: Updated test def with testcase NRFU6.5
test_definition.yaml.j2 and master_def.yaml : Updated J2 with testcase NRFU6.5

# Any specific logic/part of code you need extra attention on

Verified  power supplies fan status.

# Include the Issue number and link
https://github.com/aristanetworks/vane/issues/444

# List/Attach any dependencies/past issues that are required for this change/provide context

# Type of change

- - [ ] Bug fix (non-breaking change which fixes an issue)
- - [ ] New feature (non-breaking change which adds functionality)
- - [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- - [ ] This change requires a documentation update
- - [x] Other (please specify)

# Effort required on reviewers end

- - [ ] Easy
- - [x] Medium
- - [ ] Hard 

# How Has This Been Tested?
Unit tests and report :
**Reports:**
**With J2**
Pass case: For all fans, status is ok and speed is stable.
[with j2.docx](https://github.com/aristanetworks/vane/files/12607446/with.j2.docx)

Fail case: Status of fan is not ok or speed is not stable.(Hardcoded)
[report_2309051405.docx](https://github.com/aristanetworks/vane/files/12521246/report_2309051405.docx)

Assert: Power supply slot details are not found in the output.
[report_2308312134.docx](https://github.com/aristanetworks/vane/files/12492175/report_2308312134.docx)

Assert: Fan tray slot details are not found in the output.
[report_2308312135.docx](https://github.com/aristanetworks/vane/files/12492184/report_2308312135.docx)

Skip: Skipping testcase if device is vEOS.
[report_2308291220.docx](https://github.com/aristanetworks/vane/files/12492198/report_2308291220.docx)

HTML Reports:
[444_reports.zip](https://github.com/aristanetworks/vane/files/12607458/444_reports.zip)

**Without J2**
[pass_case_without_J2.zip](https://github.com/aristanetworks/vane/files/12607454/pass_case_without_J2.zip)

[reports_after_addressing_comments.zip](https://github.com/aristanetworks/vane/files/12735403/reports_after_addressing_comments.zip)

## New feature
    
# CI pipeline result

- - [ ] Pass
- - [ ] Fail
  
  If it fails, explain the reason and whether or not we should ignore the failure
  
# Verify Documentation Update

    
# Additional comments

### CODE REVIEW

File: nrfu_tests/test_hardware_fan_status.py [hunk 1]
```diff
@@ -0,0 +1,146 @@
+# Copyright (c) 2022 Arista Networks, Inc.  All rights reserved.
+# Arista Networks, Inc. Confidential and Proprietary.
+
+"""
+Testcases for verification of fan status.
```
Comment (hsisalgslab) [inline nrfu_tests/test_hardware_fan_status.py:5]:
```suggestion
Testcases for verification of system hardware fan status.
```

File: nrfu_tests/test_hardware_fan_status.py [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+from vane.config import dut_objs, test_defs
+from vane import tests_tools
+
+TEST_SUITE = "nrfu_test"
+
+
+@pytest.mark.nrfu_test
+class SystemHardwareFanStatusTests:
+    """
+    Testcases for verification of fan status.
```
Comment (hsisalgslab) [inline nrfu_tests/test_hardware_fan_status.py:20]: update as per above comment

File: nrfu_tests/test_hardware_fan_status.py [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+import pytest
+from pyeapi.eapilib import EapiError
+from vane.logger import logger
+from vane.config import dut_objs, test_defs
+from vane import tests_tools
+
+TEST_SUITE = "nrfu_test"
+
+
+@pytest.mark.nrfu_test
```
Comment (hsisalgslab) [inline nrfu_tests/test_hardware_fan_status.py:17]:
```suggestion
@pytest.mark.nrfu_test
@pytest.mark.system
```
need to add this marker as well

File: nrfu_tests/test_hardware_fan_status.py [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+    """
+
+    dut_parameters = tests_tools.parametrize_duts(TEST_SUITE, test_defs, dut_objs)
+    test_duts = dut_parameters["test_system_hardware_fan_status"]["duts"]
+    test_ids = dut_parameters["test_system_hardware_fan_status"]["ids"]
+
+    @pytest.mark.parametrize("dut", test_duts, ids=test_ids)
+    def test_system_hardware_fan_status(self, dut, tests_definitions):
+        """
+        TD: Testcases for verification of fan status.
```
Comment (hsisalgslab) [inline nrfu_tests/test_hardware_fan_status.py:30]:
```suggestion
        TD: Testcase for verification system hardware fan status.
```

File: nrfu_tests/test_hardware_fan_status.py [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+    """
+    Testcases for verification of fan status.
+    """
+
+    dut_parameters = tests_tools.parametrize_duts(TEST_SUITE, test_defs, dut_objs)
+    test_duts = dut_parameters["test_system_hardware_fan_status"]["duts"]
+    test_ids = dut_parameters["test_system_hardware_fan_status"]["ids"]
+
+    @pytest.mark.parametrize("dut", test_duts, ids=test_ids)
+    def test_system_hardware_fan_status(self, dut, tests_definitions):
```
Comment (hsisalgslab) [inline nrfu_tests/test_hardware_fan_status.py:28]: pls update python file name as "test_system_hardware_fan_status.py"

File: nrfu_tests/test_hardware_fan_status.py [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+        tops.output_msg = "For all fans, status is ok and speed is stable."
+
+        try:
+            """
+            TS: Running `show version` commands on DUT.
+            Verifying the 'vEOS' device is not present in output.
+            """
+            show_version_command = "show version"
+            output = dut["output"][show_version_command]["json"]
+            test_params = tops.test_parameters
```
Comment (hsisalgslab) [inline nrfu_tests/test_hardware_fan_status.py:51]: initialize it before try block

File: nrfu_tests/test_hardware_fan_status.py [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+            logger.info(
+                "On device %s: Output of %s command is: \n%s\n",
```
Comment (hsisalgslab) [inline nrfu_tests/test_hardware_fan_status.py:53]:
```suggestion
                "On device %s, Output of %s command is: \n%s\n",
```

File: nrfu_tests/test_hardware_fan_status.py [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+                f"on device {tops.dut_name}, output of {show_version_command} is:\n{output}"
+            )
+            model_name = output.get("modelName")
+
+            # Skipping testcase if device is vEOS.
+            if "vEOS" in model_name:
+                pytest.skip(f"{tops.dut_name} is vEOS device, hence test skipped.")
+
+            """
+            TS: Running `show system environment cooling detail` command and collecting
```
Comment (hsisalgslab) [inline nrfu_tests/test_hardware_fan_status.py:68]:
```suggestion
            TS: Running `show system environment cooling detail` command and verifying
```

File: nrfu_tests/test_definition.yaml [hunk 9]
```diff
@@ -0,0 +1,15 @@
+- name: nrfu_test
+  testcases:
+  - name: test_system_hardware_fan_status
+    test_id: NRFU6.5
+    description: Testcases for verification of fan status.
+    show_cmd: show system environment cooling detail
+    expected_output: null # Forming expected output dictinary for all power supplies in test.
+    power_slots: # Fan power Slots.
+      - powerSupplySlots
+      - fanTraySlots
```
Comment (hsisalgslab) [inline nrfu_tests/test_definition.yaml:10]: i guess dont need to add it in test def ..directly can parse from py file.

File: nrfu_tests/test_hardware_fan_status.py [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+                        tops.expected_output["fan_status_details"][power_slots].update(
+                            {
+                                fan_data["label"]: {
+                                    "fan_status": "ok",
+                                    "speed_stability_status": True,
+                                }
+                            }
+                        )
+
+                        tops.actual_output["fan_status_details"][power_slots].update(
```
Comment (hsisalgslab) [inline nrfu_tests/test_hardware_fan_status.py:105]: image
we dont use camel casing for output dict keys

File: nrfu_tests/test_hardware_fan_status.py [hunk 11]
```diff
@@ -0,0 +1,146 @@
...
+                    if slot_details != tops.actual_output.get("fan_status_details").get(slots):
+                        tops.output_msg += f"For {slots}:\n"
+                    if notok_fans[slots]:
+                        fan_label = ", ".join(notok_fans[slots])
+                        tops.output_msg += f"State of following fans is not ok: {fan_label}\n"
+                    if unstable_fans[slots]:
+                        fan_label = ", ".join(unstable_fans[slots])
+                        tops.output_msg += (
+                            "Speed of following fans is not stable, there speed is greater than"
+                            f" 80: {fan_label}\n\n"
```
Comment (hsisalgslab) [inline nrfu_tests/test_hardware_fan_status.py:132]: 80 can be taken from test def ?

File: nrfu_tests/test_hardware_fan_status.py (hunk 1 above)
Comment (hsisalgslab) [reply nrfu_tests/test_hardware_fan_status.py]: This is not addressed

**Review by hsisalgslab — APPROVED** [review]

File: nrfu_tests/test_system_hardware_fan_status.py [hunk 12]
```diff
@@ -0,0 +1,152 @@
...
+        """
+        TD: Testcase for verification of system hardware fan status.
+        Args:
+            dut(dict): details related to a particular DUT
+            tests_definitions(dict): test suite and test case parameters.
+        """
+        tops = tests_tools.TestOps(tests_definitions, TEST_SUITE, dut)
+        test_params = tops.test_parameters
+        tops.expected_output = {"fan_status_details": {}}
+        tops.actual_output = {"fan_status_details": {}}
```
Comment (vishwajeetgslab) [inline nrfu_tests/test_system_hardware_fan_status.py:40]: can we add it as {fans_slots:  {fan_tray_slots: {}, power_supply_slots:{}}

File: nrfu_tests/test_system_hardware_fan_status.py [hunk 13]
```diff
@@ -0,0 +1,152 @@
...
+        unstable_fans, notok_fans = {}, {}
+
+        # forming output message if test result is passed
+        tops.output_msg = "For all fans, status is ok and speed is stable."
+
+        try:
+            """
+            TS: Running `show version` commands on DUT.
+            Verifying the 'vEOS' device is not present in output.
+            """
```
Comment (vishwajeetgslab) [inline nrfu_tests/test_system_hardware_fan_status.py:51]: """
TS: Running `show version` commands on the device and skipping the test case if device platform is 'vEOS'.
 """

File: nrfu_tests/test_system_hardware_fan_status.py [hunk 14]
```diff
@@ -0,0 +1,152 @@
...
+                        tops.output_msg += f"For {slots}:\n"
+                    if notok_fans[slots]:
+                        fan_label = ", ".join(notok_fans[slots])
+                        tops.output_msg += f"State of following fans is not ok: {fan_label}.\n"
+                    if unstable_fans[slots]:
+                        fan_label = ", ".join(unstable_fans[slots])
+                        tops.output_msg += (
+                            f"Speed of following fans is not stable: {fan_label}, and their speed"
+                            f" is greater than {test_params['fan_speed_limit']}.\n\n"
+                        )
```
Comment (vishwajeetgslab) [inline nrfu_tests/test_system_hardware_fan_status.py:139]: We need to add the current state of fans and speed as well in observation section

**Review by vishwajeetgslab — CHANGES_REQUESTED** [review]

File: nrfu_tests/test_definition.yaml [hunk 15]
```diff
@@ -0,0 +1,12 @@
+- name: nrfu_tests
+  testcases:
+  - name: test_system_hardware_fan_status
```
Comment (vishwajeetgslab) [inline nrfu_tests/test_definition.yaml:3]: format properly

**Review by vishwajeetgslab — CHANGES_REQUESTED** [review]

**Review by vishwajeetgslab — APPROVED** [review]

File: nrfu_tests/test_system_hardware_fan_status.py [hunk 16]
```diff
@@ -0,0 +1,155 @@
...
+                        fan_speed_details.update({fan_data["label"]: fan_data["configuredSpeed"]})
+
+            # Forming output message if the test result is failed.
+            if tops.expected_output != tops.actual_output:
+                tops.output_msg = "\nFollowing power supply slots, fans are in erroneous state :\n"
+                for slots, slot_details in tops.expected_output["fans_slots"].items():
+                    if slot_details != tops.actual_output["fans_slots"].get(slots):
+                        tops.output_msg += f"For {slots.replace('_', ' ')}:\n"
+                        for fan_lable, fan_details in slot_details.items():
```
Comment (kgrozis-arista) [inline nrfu_tests/test_system_hardware_fan_status.py:121]: Should fan_lable be fan_label?

File: nrfu_tests/test_system_hardware_fan_status.py [hunk 17]
```diff
@@ -0,0 +1,155 @@
...
+                            for fan_status_key, fan_status_value in fan_details.items():
+                                actual_fan_status = (
+                                    tops.actual_output["fans_slots"]
+                                    .get(slots)
+                                    .get(fan_lable)
```
Comment (kgrozis-arista) [inline nrfu_tests/test_system_hardware_fan_status.py:126]: Should fan_lable be fan_label?

File: nrfu_tests/test_system_hardware_fan_status.py [hunk 18]
```diff
@@ -0,0 +1,155 @@
...
+                                )
+                                if fan_status_value != actual_fan_status:
+                                    if actual_fan_status is None:
+                                        tops.output_msg += f"{fan_lable}: Fan status is 'Not ok'.\n"
+                                    else:
+                                        if fan_status_key == "speed_stability_status" and (
+                                            int(fan_speed_details[fan_lable]) > 80
+                                        ):
+                                            tops.output_msg += (
+                                                f"{fan_lable}: Fan status is 'ok', however current"
```
Comment (kgrozis-arista) [inline nrfu_tests/test_system_hardware_fan_status.py:137]: Should fan_lable be fan_label?

File: nrfu_tests/test_system_hardware_fan_status.py [hunk 19]
```diff
@@ -0,0 +1,155 @@
...
+                                    .get(slots)
+                                    .get(fan_lable)
+                                    .get(fan_status_key)
+                                )
+                                if fan_status_value != actual_fan_status:
+                                    if actual_fan_status is None:
+                                        tops.output_msg += f"{fan_lable}: Fan status is 'Not ok'.\n"
+                                    else:
+                                        if fan_status_key == "speed_stability_status" and (
+                                            int(fan_speed_details[fan_lable]) > 80
```
Comment (kgrozis-arista) [inline nrfu_tests/test_system_hardware_fan_status.py:134]: Should fan_lable be fan_label?

File: nrfu_tests/test_system_hardware_fan_status.py [hunk 20]
```diff
@@ -0,0 +1,155 @@
...
+                            for fan_status_key, fan_status_value in fan_details.items():
+                                actual_fan_status = (
+                                    tops.actual_output["fans_slots"]
+                                    .get(slots)
+                                    .get(fan_lable)
+                                    .get(fan_status_key)
+                                )
+                                if fan_status_value != actual_fan_status:
+                                    if actual_fan_status is None:
+                                        tops.output_msg += f"{fan_lable}: Fan status is 'Not ok'.\n"
```
Comment (kgrozis-arista) [inline nrfu_tests/test_system_hardware_fan_status.py:131]: Should fan_lable be fan_label?

File: nrfu_tests/test_system_hardware_fan_status.py [hunk 21]
```diff
@@ -0,0 +1,155 @@
...
+                                    else:
+                                        if fan_status_key == "speed_stability_status" and (
+                                            int(fan_speed_details[fan_lable]) > 80
+                                        ):
+                                            tops.output_msg += (
+                                                f"{fan_lable}: Fan status is 'ok', however current"
+                                                f" fan speed '{fan_speed_details.get(fan_lable)}'"
```
Comment (kgrozis-arista) [inline nrfu_tests/test_system_hardware_fan_status.py:138]: Should fan_lable be fan_label?

**Review by kgrozis-arista — CHANGES_REQUESTED** [review]

File: nrfu_tests/test_system_hardware_fan_status.py (hunk 21 above)
Comment (rohitsinghgslab) [reply nrfu_tests/test_system_hardware_fan_status.py]: Done

File: nrfu_tests/test_system_hardware_fan_status.py (hunk 19 above)
Comment (rohitsinghgslab) [reply nrfu_tests/test_system_hardware_fan_status.py]: Done

File: nrfu_tests/test_system_hardware_fan_status.py (hunk 18 above)
Comment (rohitsinghgslab) [reply nrfu_tests/test_system_hardware_fan_status.py]: Done

File: nrfu_tests/test_system_hardware_fan_status.py (hunk 17 above)
Comment (rohitsinghgslab) [reply nrfu_tests/test_system_hardware_fan_status.py]: Done

File: nrfu_tests/test_system_hardware_fan_status.py (hunk 16 above)
Comment (rohitsinghgslab) [reply nrfu_tests/test_system_hardware_fan_status.py]: Done, thanks

Comment (kgrozis-arista) [reply nrfu_tests/test_system_hardware_fan_status.py]: lgtm

File: nrfu_tests/test_system_hardware_fan_status.py (hunk 18 above)
Comment (kgrozis-arista) [reply nrfu_tests/test_system_hardware_fan_status.py]: lgtm

File: nrfu_tests/test_system_hardware_fan_status.py (hunk 20 above)
Comment (kgrozis-arista) [reply nrfu_tests/test_system_hardware_fan_status.py]: lgtm

**Review by kgrozis-arista — APPROVED** [review]

**Pull request merged**